### PR TITLE
Implement JSON Blob storage as a module

### PR DESCRIPTION
**Pull request by Wedjaa**
Since jsonblob has been having some problems lately - mainly with their certificates - we developed a custom backend for saving the JSON configuration blobs. These changes to Mirador allow the use of a custom module to save the configuration, by default it will still use jsonblob. 

Our module is not included in this PR because it's on a custom elasticsearch based backend, but I can push it as a separate module with the relative node.js routes if anyone is interested.

### CODE REVIEW

**Comment by Wedjaa** [comment]
I am thinking of making the URL to the jsonblob API - now a constant in the module implementation - a configuration parameter, allowing the deployment of a on-site jsonblob API endpoint. I don't know if anyone is interested in it.

**Comment by aeschylus** [comment]
Hi @Wedjaa. This is a really great contribution. Ultimately we did intend for people to be able to configure this, so I would love to merge it.

We aren't currently set up to automatically test remote endpoints, but do you have w working demo, and could you write a few tests for the Mirador-facing side of the API?

**Comment by Wedjaa** [comment]
Hi,

  I will try and extract the interface I created and put it out as a node.js module so that it would be easy to test in any environment - I will implement the tests as well.

**Comment by Wedjaa** [comment]
Have not forgotten - just knee deep in a release sprint - will work on
this once I get some time!

2015-06-25 13:56 GMT-05:00 aeschylus <notifications@github.com>:

**Comment by aeschylus** [comment]
We should prioritise this in 2.1. This will affect the scaffolded wiki page as well. Many have been asking how to enable bookmarking and we haven't explained it at all in the documentation.

**Comment by nicolasfranck** [comment]
This would indeed be a very good enhancement!
That way one can setup its own jsonblob-server,
and make sure it works instead of depending on 
someone else's service.

Apparently this branch has merge conflicts? Anything I can do?

**Comment by regisrob** [comment]
+1. I plan to set up my own jsonblob-server too (at least as a first test)

**Comment by sdellis** [comment]
I haven't heard about this until now, and couldn't really find any documentation on the intent.  I'm curious, what is this jsonblob-server intended to store?  Configuration objects, annotations, manifests/collections?  And this is only for writing tests?

**Comment by regisrob** [comment]
AFAIK the purpose is only to store configuration objects remotely (see https://github.com/IIIF/mirador/issues/559#issuecomment-121037276)

**Comment by Wedjaa** [comment]
Aligned with the current master branch.

**Comment by rsinghal** [comment]
@Wedjaa I like what you did with the code.  I am having one issue - I am getting 406 errors from jsonblob (and was not getting it from current master).  Looking at the Headers tab in Chrome Developers Tools I am seeing 
PR:
```
Request Payload
[object Object]
```

vs master:
```
Request Payload
{"workspaceType":"singleObject","saveSession":true,"workspaces":{"singleObject":{"label":"Single Object","addNew":false,"move":false,"iconClass":"image"},"compare":{"label":"Compare","iconClass":"columns"},"bookReading":{"defaultWindowOptions":{},"label":"Book Reading","addNew":true,"move":false,"iconClass":"book"}},"layout":"{\"type\":\"row\",\"depth\":0,\"value\":0,\"x\":3,\"y\":3,\"dx\":741,\"dy\":691,\"address\":\"row1\",\"id\":\"5a17ae9e-1d14-482b-b28e-12923603d77c\"}","windowObjects":[{"loadedManifest":"https://oculus.harvardx.harvard.edu/manifests/vjx:201501","id":"e5b9f4ba-a346-4d59-acd1-622d4a8bda1d","viewType":"ThumbnailsView","canvasID":"https://oculus.harvardx.harvard.edu/manifests/vjx:201501/canvas/canvas-40_000a_screen_top.json","imageMode":"ImageView","slotAddress":"row1","bottomPanelVisible":false,"sidePanelVisible":true}],"defaultWindowSettings":{},"availableAnnotationModes":[],"availableAnnotationDrawingTools":[],"availableCanvasTools":[],"mainMenuSettings":{"show":true,"buttons":{"bookmark":true,"layout":true,"options":false,"fullScreenViewer":true}},"workspacePanelSettings":{"maxRows":5,"maxColumns":5,"preserveWindows":true},"showAddFromURLBox":true,"buildPath":"build/mirador/","i18nPath":"locales/","imagesPath":"images/","logosPath":"images/logos/","repoImages":{"other":"iiif_logo.png"},"annotationEndpoint":{},"sharingEndpoint":{"url":"","storeId":123,"APIKey":"23983hf98j3f9283jf2983fj"},"lockController":{"lockProfile":"lazyZoom","notifyMaxMin":true},"id":"viewer","data":[{"manifestUri":"https://oculus.harvardx.harvard.edu/manifests/vjx:201501","location":"Harvard University"}]}
```

**Comment by beaudet** [comment]
Are there suggested best practices to making sure this is compatible across Mirador versions?

**Comment by Wedjaa** [comment]
@rsinghal Sorry - I may have introduced the bug while rebasing on the new master. Fixing it.

**Comment by Wedjaa** [comment]
@rsinghal - bugs introduced by me; fixed now. Tested fine. Preparing a node based server for and in-house jsonblob-like service with a Redis backend.

File: js/src/utils/jsonBlobApi.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  $.JSONBlobAPI.prototype = {
+	readSync: function(blobId) {
+		return syncRequest('GET', blobId, this.options.ssl);
+	},
+	save: function(blob) {
+		var promise = jQuery.Deferred().promise();
+		asyncRequest('POST', '', this.options.ssl, blob, function(data, textStatus, request) {
+			var blobid = request.getResponseHeader('X-Jsonblob');
+			promise.resolve(blobid);
```
Comment (rsinghal) [inline js/src/utils/jsonBlobApi.js]: @Wedjaa Thanks, now I'm getting a ```Uncaught TypeError: promise.resolve is not a function``` here.  Seems weird - do we have different versions of jQuery?

**Comment by Wedjaa** [comment]
@rsinghal That's very strange -- I'm testing with the default index.html from master, will have to check what jquery is in that.

**Comment by Wedjaa** [comment]
@rsinghal The jquery used by mirador is 1.9.1 - it should support promises. I have tested and it's working on Chrome. What browser are you testing in? We may be hitting an issue with jquery implementation of promises.

**Comment by Wedjaa** [comment]
@rsinghal - OK, there is a bug in the implementation of promises - but it's the bug making it work for me! :) Will fix and update the pull request. Funnily enough not even phantomjs seems to pick up on this error: .resolve() should not be exposed on the deferred.promise() - but it is on some JS engines. The fix is using deferred as they are supposed to be used.

**Comment by rsinghal** [comment]
@Wedjaa Glad you figured it out!  I am also using Chrome, so I don't know why it cropped up for me.

**Comment by Wedjaa** [comment]
@rsinghal Using the promises correctly, for a change. Changed also the *jsonBlob* verbiage to *blobId*, since it's really what we are passing around. Let me know if you see other issues. Today I'll be travelling - but I hope to release the Redis backed jsonBlob like server sometime this week.

**Comment by rsinghal** [comment]
@Wedjaa it's working for me.  Merging now.

**Comment by leanderseige** [comment]
How can I configure mirador to use an own jsonblob-server?
I installed https://github.com/tburch/jsonblob and tried (with no success):
```javascript
jsonStorageEndpoint: {
                        name: 'JSONBlob API Endpoint',
                        module: 'JSONBlobAPI',
                        options: {
                                ssl: true,
                                url: 'http://127.0.0.1:38080/api/jsonBlob'
                        }
                }
```

**Comment by Wedjaa** [comment]
@leanderseige I haven't tried it with the jsonblob server you are pointing to. Do you have it running somewhere where I can test it?

**Comment by rsinghal** [comment]
@Wedjaa I think @leanderseige set up his own instance of the jsonblob (that github repo is what runs jsonblob.com) and wants to change JSONBLOB_API_BASE to point to a different URL.  Can you update the settings to allow for that?

**Comment by Wedjaa** [comment]
@rsinghal Installed JSONBlob locally   - working on the settings.

**Comment by leanderseige** [comment]
Thank you, @Wedjaa and @rsinghal!

**Comment by Wedjaa** [comment]
@leanderseige and @rsinghal -- Created PR #714 to allow the configuration of the JSONBlob plugin.

**Comment by beaudet** [comment]
I still think it would advantageous to embed a Mirador version number in the data so that future versions of Mirador can provide backwards compatibility in case external entities decide to persist this data.  If the intention is that this data should be temporal only, we should document that fact and encourage persisting implementations to handle the versioning problem themselves.

**Pull request merged**